core-backend/common/src/funcs.rs

```diff
@@ -174,14 +174,19 @@ where
 
     #[host(fallible, cost = RuntimeCosts::Read, err_len = LengthBytes)]
     pub fn read(ctx: &mut R, at: u32, len: u32, buffer_ptr: u32) -> Result<(), R::Error> {
-        let (buffer, mut gas_left) = ctx.ext_mut().read(at, len)?;
-        let buffer = buffer.to_vec();
-
-        let write_buffer = ctx.register_write(buffer_ptr, len);
-        ctx.memory_manager_write(write_buffer, &buffer, &mut gas_left)?;
-
-        ctx.ext_mut().set_gas_left(gas_left);
-        Ok(())
+        let payload_holder = ctx.ext_mut().lend_payload(at, len)?;
+        payload_holder
+            .use_with_job::<MemoryAccessError, _>(|payload_to_slice| {
+                let write_buffer = ctx.register_write(buffer_ptr, len);
+                let write_res = ctx.write(write_buffer, payload_to_slice.to_slice());
+                let reclaim_res = ctx
+                    .ext_mut()
+                    .reclaim_payload(payload_to_slice.into_holder());
```

Comment:
```suggestion
                 let mut aa = MessageContext::new(Default::default(), Default::default(), ContextSettings::new(0, 0, 0, 0, 0, 0));
                let mut bb = PayloadSliceHolder::try_new((0, 0), &mut aa).unwrap();
                let reclaim_res = ReclaimBoundResult::from((&mut aa, &mut bb));
```
This code is also compiles well but obviously incorrect.

Reply:
Lol, interesting  but tests goes well))) may be need to add double read test

Reply:
Definitely need this test

Reply:
> ```suggestion
> 
>                  let mut aa = MessageContext::new(Default::default(), Default::default(), ContextSettings::new(0, 0, 0, 0, 0, 0));
> 
>                 let mut bb = PayloadSliceHolder::try_new((0, 0), &mut aa).unwrap();
> 
>                 let reclaim_res = ReclaimBoundResult::from((&mut aa, &mut bb));
> 
> ```
> 
> This code is also compiles well but obviously incorrect.

There's no defence against writing such wrong stuff at all. It violates everything. It's not the reason that current implementation is unsafe. We can do the same stuff in every syscall implementation. To have a proper security against that, we must make some types private.

Reply:
Yes, but what the reason in such complex realisation if it does not guaranty compilation time safety. Better to use take_payload() and set_payload() carefully. You can also create PayloadHandler with panic in drop. Or there is also solution for compilation time, but it is a little strange maybe =) this one https://stackoverflow.com/questions/71694803/rust-type-that-requires-manual-drop

Reply:
> Yes, but what the reason in such complex realisation

1. That's the price of introducing type-safe approach. You literally create types for "every" step. As a result, you get code safe from **unintentional** wrong usage (when you forget to call something or use some type and etc.), which will error in compile time.

>  if it does not guaranty compilation time safety.

2. I guess you misunderstand what is a compile time safety feature. Basically, you use types to **thoroughly describe** your business logic. Wrong usage of types will mean a wrong implementation of business logic and you will understand that either from compilation error (which is the best case) or from tests (which is good too, but worse). What you have done in comment example, is that you have abused fully the business logic of `read`. Following this practise I can abuse business logic of every sys-call and then say, that implementation is not so safe, so why to make any "_complex realisation_".
3. Current type-safe implementation guarantees no **unintentional** errors, when you forget things like here https://github.com/gear-tech/gear/commit/c1d951afadeebc9f9490661ea04fb35af2122977 (return from funcs before setting gas), or like calling `to_vec` in `read` here https://github.com/gear-tech/gear/commit/c1d951afadeebc9f9490661ea04fb35af2122977, or like we had before refactoring the `wasmi` situations when termination reason caused gas counter not being updated. These are **unintentional** errors, because developer "forgets" doing something. They were in our code base for a long time. We can't rely much on reviews when there are vague errors from too much implicit business logic not guarded by types. But, again, example you provided is a full abuse of business logic: you can't and not expected to create a `MessageContext` there, it's unarguably wrong.
4. You can make a fair point that `MessageContext` then mustn't be instantiated so easy with just call to `new` from anywhere. Yes, it must be possible only to create it from `execute_wasm` method. I decided not to implement this constraint as it was a bit too much, because no-one with a common sense will think to commit to `master` such a code, that creates an additional `MessageContext` in a wrong place, when this type is hidden under `ctx` param of the function.

> Better to use take_payload() and set_payload() carefully.

5. Again, "carefully" is a wrong approach now. I gave you 3 examples of code in "master". Were dev not careful during that? I guess they were. Mistakes happen. But not guarantee they won't happen further. But with strong types we get more guarantees for a safer usage of externalities methods.

> You can also create PayloadHandler with panic in drop

7. Implementing such runtime approach or, for example, returning error from `into_ext_info` if payload was taken and not set again, will require returning a `System` error from backend, which will be considered as a panic on `pallet_gear`.

> Or there is also solution for compilation time, but it is a little strange maybe =) this one https://stackoverflow.com/questions/71694803/rust-type-that-requires-manual-drop

`ManuallyDrop` brings associated risks. It's actually making solution more complex. Complexity is not with 3-4 more additional types, but with what they do. Do you find, what introduced in this PR types do more risky and complex than `ManuallyDrop`?